cache/keytemplate/keytemplate.go

```diff
@@ -0,0 +1,87 @@
+package keytemplate
+
+import (
+	"bytes"
+	"fmt"
+	"runtime"
+	"text/template"
+
+	"github.com/bitrise-io/go-utils/v2/env"
+	"github.com/bitrise-io/go-utils/v2/log"
+)
+
+// Model ...
+type Model struct {
+	envRepo env.Repository
+	logger  log.Logger
+	os      string
+	arch    string
+}
+
+// BuildContext contains metadata about the build that gets exposed to the template
+type BuildContext struct {
+	Workflow   string
+	Branch     string
+	CommitHash string
+}
+
+type templateInventory struct {
+	OS         string
+	Arch       string
+	Workflow   string
+	Branch     string
+	CommitHash string
+}
+
+// NewModel ...
+func NewModel(envRepo env.Repository, logger log.Logger) Model {
+	return Model{
+		envRepo: envRepo,
+		logger:  logger,
+		os:      runtime.GOOS,
+		arch:    runtime.GOARCH,
+	}
+}
+
+// Evaluate returns the final string from a key template and the provided build context
+func (m Model) Evaluate(key string, buildContext BuildContext) (string, error) {
+	funcMap := template.FuncMap{
+		"getenv": m.getEnvVar,
+	}
+
+	tmpl, err := template.New("").Funcs(funcMap).Parse(key)
+	if err != nil {
+		return "", fmt.Errorf("invalid template: %w", err)
+	}
+
+	inventory := templateInventory{
+		OS:         m.os,
+		Arch:       m.arch,
+		Workflow:   buildContext.Workflow,
+		Branch:     buildContext.Branch,
+		CommitHash: buildContext.CommitHash,
+	}
+	m.validateInventory(inventory)
+
+	resultBuffer := bytes.Buffer{}
+	if err := tmpl.Execute(&resultBuffer, inventory); err != nil {
+		return "", err
+	}
+	return resultBuffer.String(), nil
+}
+
+func (m Model) getEnvVar(key string) string {
+	return m.envRepo.Get(key)
```

Comment:
NOTE: It could make sense to add a warning if an env is not set.

Reply:
Great idea, fixed it.